pkg/binding/bindingpolicy-resolver.go

```diff
@@ -66,26 +65,29 @@ type BindingPolicyResolver interface {
 	// singleton status reporting requirement in the resolution.
 	NoteBindingPolicy(bindingpolicy *v1alpha1.BindingPolicy)
 
-	// NoteObject updates the maintained bindingpolicy's objects resolution for the
-	// given bindingpolicy key. If the object is being deleted, it is removed from
-	// the resolution if exists.
+	// EnsureObjectIdentifierWithVersion ensures that an object's identifier is
+	// in the resolution for the given bindingpolicy key, and is associated
+	// with the given resourceVersion.
 	//
 	// The returned bool indicates whether the bindingpolicy resolution was changed.
 	// If no resolution is associated with the given key, an error is returned.
-	NoteObject(bindingPolicyKey string, obj runtime.Object) (bool, error)
-	// RemoveObject removes the given object from the maintained bindingpolicy's
-	// objects resolution for the given bindingpolicy key.
+	EnsureObjectIdentifierWithVersion(bindingPolicyKey string, objIdentifier util.ObjectIdentifier,
+		resourceVersion string) (bool, error)
+	// RemoveObjectIdentifier removes the given object identifier from the
+	// resolution for the given bindingpolicy key.
 	//
 	// The returned bool indicates whether the bindingpolicy resolution was changed.
-	RemoveObject(bindingPolicyKey string, obj runtime.Object) bool
+	// If no resolution is associated with the given key, false is returned.
+	RemoveObjectIdentifier(bindingPolicyKey string, objIdentifier util.ObjectIdentifier) bool
+	// GetObjectIdentifiers returns the object identifiers associated with the
+	// given bindingpolicy key.
+	// If no resolution is associated with the given key, an error is returned.
+	GetObjectIdentifiers(bindingPolicyKey string) (sets.Set[util.ObjectIdentifier], error)
+
 	// SetDestinations updates the maintained bindingpolicy's
 	// destinations resolution for the given bindingpolicy key.
 	// The given destinations set is expected not to be mutated after this call.
 	SetDestinations(bindingPolicyKey string, destinations sets.Set[string])
```

Comment:
You're right, `SetDestinations` does not return an error because it is always called after `NoteBindingPolicy`, but that should not impact the design of the resolver. 

Updated.